### PR TITLE
Enabled a consistent variable ordering in THM from components

### LINKED ISSUE #28503: Use a consistent variable ordering in THM
### Motivation

After changing some variable naming in THM, it was discovered that the variable ordering in the nonlinear system ultimately depends on the variable names themselves, since in `Simulation` there is a loop over an unordered map keyed from variable name to add the variables to the problem. The issue with changing variable ordering is that it affects the convergence of linear solvers, so changing a variable's name can cause divergence for example.

### Design

This would use a standardized variable ordering in THM. The components in THM have a limited set of variables, such as `rhoA`, `rhouA`, `rhoEA`, `T_solid`, `rhoV`, etc. We could keep a list of these variables in the order we desire and then sort the variable data by this ordering. If there happen to be variables not in this list, they can be at the end of the list in alphabetical order.

### Impact

This will likely cause some convergence differences, hopefully good ones.

### PR DESCRIPTION
**Pull request by joshuahansel**
Closes #28503

See the additions to the Components system page for a description of what this PR does. I'll include a link after documentation builds.

### CODE REVIEW

File: modules/thermal_hydraulics/src/base/Simulation.C [hunk 1]
```diff
@@ -564,6 +572,63 @@ Simulation::addComponentScalarIC(const VariableName & var_name, const std::vecto
...
+  std::vector<std::pair<VariableName, int>> registered_var_index_pairs;
+  for (const auto & var_and_index : _component_variable_order_map)
+  {
+    registered_var_index_pairs.push_back(var_and_index);
+
+    const auto ind = var_and_index.second;
+    if (std::find(indices.begin(), indices.end(), ind) != indices.end())
+      mooseError("The index ", ind, " was used for multiple component variables.");
+    else
+      indices.push_back(ind);
```
Comment (GiudGiud) [inline modules/thermal_hydraulics/src/base/Simulation.C:590]: if you used a set for indices you could check whether insertion worked. a little faster (N log N)
N searches in vector is N2

fine for small numbers here

File: modules/thermal_hydraulics/src/base/Simulation.C [hunk 2]
```diff
@@ -564,6 +572,63 @@ Simulation::addComponentScalarIC(const VariableName & var_name, const std::vecto
...
+            { return a.second < b.second; });
+
+  // Loop over the ordered, registered variable names and add a variable to the
+  // sorted list if in vars_unsorted. When this happens, delete the element from
+  // vars_unsorted, leaving only unregistered variable names after the loop.
+  std::vector<VariableName> vars_sorted;
+  for (const auto & var_index_pair : registered_var_index_pairs)
+  {
+    const auto & var = var_index_pair.first;
+    if (std::find(vars_unsorted.begin(), vars_unsorted.end(), var) != vars_unsorted.end())
```
Comment (GiudGiud) [inline modules/thermal_hydraulics/src/base/Simulation.C:616]: if there's no duplicates do you really need to search? you can always insert in vars_sorted and delete in vars_unsorted

Comment (GiudGiud) [inline modules/thermal_hydraulics/src/base/Simulation.C:616]: if there are no duplicates do you really need to search? you can always insert in vars_sorted and delete in vars_unsorted

File: modules/thermal_hydraulics/src/base/Simulation.C [hunk 3]
```diff
@@ -582,15 +647,27 @@ Simulation::addVariables()
...
   for (auto && comp : _components)
     comp->addVariables();
 
-  // pass the variables to MOOSE
-  for (auto && v : _vars)
+  // Sort the variables for a consistent ordering
+  const auto var_names = sortAddedComponentVariables();
+
+  // Report the ordering if the executioner is verbose
+  if (_thm_app.getExecutioner()->getParam<bool>("verbose"))
```
Comment (GiudGiud) [inline modules/thermal_hydraulics/src/base/Simulation.C:658]: use verbose_setup parameter instead, it s a problem parameter and it fits here

File: modules/thermal_hydraulics/src/base/Simulation.C (hunk 2 above)
Comment (joshuahansel) [reply modules/thermal_hydraulics/src/base/Simulation.C:616]: I don't understand the suggestion. `vars_sorted` should only contain the vars that were in `vars_unsorted`, so the find determines this.

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**